File: leetcode/leetcode_cpp/reverse-words-in-a-string.cpp
#include <iostream>
#include <cassert>

#include <sstream>

using namespace std;

/**
https://leetcode.com/problems/reverse-words-in-a-string

Solution1. iterate from begin to end
time: o(n).
space: o(n)

Solution2_1. reverse words with erase
time: o(n)
space: o(1)

1. erase spaces between words
2. reverse words. and add space
3. reverse all string

Solution2_2. reverse words with resize
time: o(n)
space: o(1)

1. 
2. reverse words. and add space
3. reverse all string
*/

class Solution {
public:
    string reverseWords_1(string& s) {
        string res;
        string word;
        for (int i=0; i<=s.size(); ++i) {
            if (i == s.size() || isspace(s[i])) {
                if (!word.empty()) res = word + " " + res;
                word = "";
            }
            else {
                word += s[i];
            }
            
        }
        if (res.size()) res.pop_back();
        return res;
    }

    string reverseWords_2_1(string& s) {
        int l = 0, r = 0;
        while (r<s.size()) {
            while (r<s.size() && isspace(s[r])) r++;
            if (l != r) s.erase(l, r-l);
            r = l;
            if (r>=s.size()) break;
            
            while (r<s.size() && !isspace(s[r])) r++;
            if (l != r) reverse(s.begin()+l, s.begin()+r);
            l = r+1;
        }
        if (!s.empty() && isspace(s[s.size() - 1])) s.pop_back();
        reverse(s.begin(), s.end());
        return s;
    }

    string reverseWords_2_2(string& s) {
        reverse(s.begin(), s.end());
        int l=0, r=0, n=s.size(), lastIdx = 0;
        while (r<n) {
            while (r<n && s[r] == ' ') r++;
            int startIdx = l;
            while (r<n && s[r] != ' ') {
                s[l++] = s[r++];
                lastIdx = l;
            }
            reverse(s.begin() + startIdx, s.begin() + lastIdx);
            s[l++] = ' ';
        }
        s.resize(lastIdx);
        return s;
    }

    string reverseWords(string s) {
        return reverseWords_2_2(s);
    }
};

int main()
{
    assert(Solution().reverseWords("  e f  g  ") == ("g f e"));
    assert(Solution().reverseWords("the sky is blue") == ("blue is sky the"));
    return 0;
}

File: leetcode/leetcode_py/reverse-words-in-a-string.py

'''
https://leetcode.com/problems/reverse-words-in-a-string

Solution1. iterate from begin to end
time: o(n).
space: o(n)

Solution2_1. reverse words with erase
time: o(n)
space: o(1)

1. erase spaces between words
2. reverse words. and add space
3. reverse all string

Solution2_2. reverse words with resize
time: o(n)
space: o(1)

1. skip empty space
2. reverse each word
3. resize string
'''


class Solution:
    def reverseWords_builtin1(self, s: str) -> str:
        res = ''
        for item in s.split():
            res = item + ' ' + res
        return res[:-1]
    
    def reverseWords_builtin2(self, s: str) -> str:
        return " ".join(s.strip().split()[::-1])
    
    def reverseWords_iterate(self, s: str) -> str:
        res, substr = '', ''
        for c in s:
            if c.isspace():
                if substr: res = substr + ' '+ res
                substr = ''
            else:
                substr += c
        if substr: res = substr + ' '+ res
        if res: res = res[:-1]
        return res
        
    def reverseWords_3(self, s: str) -> str:
        s = s[::-1]
        print(s)
        n = len(s)
        i, l, r = 0, 0, 0
        while i < n:
            while i < n and s[i] != ' ':
                s = s[:r] + s[i] + s[r+1:]
                r += 1
                i += 1
            
            if l < r:
                s = s[:l] + s[l:r][::-1] + ' ' + s[r+1:]
                r += 1
                l = r
            i += 1
        
        while r > 0 and s[r-1] == ' ': r -= 1
        return s[:r]

    def reverseWords_1(self, s: str) -> str:
        res = ''
        word = ''
        for i in range(len(s) + 1):
            if i == len(s) or s[i] == ' ':
                if word:
                    res = word + ' ' + res
                word = ''
            else:
                word += s[i]
        return res[:len(res)-1] if res else ''

    def reverseWords_2(self, s: str) -> str:
        s = list(s)
        s = s[::-1]
        l, r, n, lastIdx = 0, 0, len(s), 0
        while r < n:
            while r < n and s[r] == ' ':
                r += 1
            startIdx = l
            while r < n and s[r] != ' ':
                s[l] = s[r]
                l += 1
                r += 1
                lastIdx = l
            
            s = s[:startIdx] + s[startIdx:lastIdx][::-1] + s[lastIdx:]
            if l < n: s[l] = ' '
            l += 1
                
        return "".join(s[:lastIdx])

    def reverseWords(self, s: str) -> str:
        return self.reverseWords_2(s)
        